sc apps: Dashboard for visualizing how spread-out pods are across nodes

Zash opened the pull request:
<!-- Choose you PR title carefully as it will be used as the entry in the changelog! -->

> [!warning]
> **This is public repository, ensure not to disclose:**
> - [x] personal data beyond what is necessary for interacting with this pull request
> - [x] business confidential information, such as customer names

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [x] kind/feature       <!-- This PR adds a new feature -->
- [ ] kind/improvement   <!-- This PR changes an existing feature -->
- [ ] kind/deprecation   <!-- This PR removes an existing feature -->
- [ ] kind/documentation <!-- This PR contains documentation -->
- [ ] kind/clean-up      <!-- This PR cleans up technical debt -->
- [ ] kind/bug           <!-- This PR fixes a bug -->
- [ ] kind/other         <!-- This PR does something else -->

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change  <!-- This PR introduces an admin facing change, add "Platform Administrator notice" section -->
- [ ] kind/dev-change    <!-- This PR introduces a dev facing change, add "Application Developer notice" section -->
- [ ] kind/security      <!-- This PR introduces a critical security fix, add "Security notice" section -->
- [ ] [kind/adr]()       <!-- This PR implements an ADR, add the link -->

<!-- Uncomment the additional sections that applies. -->

<!-- Additional information to be added in the release notes
### Release notes
...
-->

<!-- Additional information with kind/admin-change
### Platform Administrator notice
...
-->

<!-- Add additional information with kind/dev-change
### Application Developer notice
...
-->

<!-- Add additional information with kind/security
### Security notice
...
-->

### What does this PR do / why do we need this PR?

<!-- Add description of the change -->
This dashboard lets you explore how well Pods are spread out across either Nodes or Zones. 

<!-- Add all issues that are fixed by this PR, use "Part of" instead of "Fixes" if you want to keep issues open. -->
- Fixes #1345

#### Additional information to reviewers

#### Screenshots

![image](https://github.com/elastisys/compliantkubernetes-apps/assets/197474/08a02b7d-86c9-48f1-87dc-a6bd8671c614)


#### Checklist

<!-- This section is not added to the changelog or release notes, it is to help you as a contributor and reviewers. -->

- [x] Proper commit message prefix on all commits
  <!-- Example of commit message prefixes:
    - all: changes to multiple areas
    - apps: changes to application running in all clusters
    - apps sc: changes to applications running in the service cluster
    - apps wc: changes to applications running in the workload cluster
    - bin: changes to management binaries or scripts
    - config: changes to configuration
    - docs: changes to documentation
    - tests: changes to tests
    - pipeline: changes to the pipeline
    - release: release related
  -->
- Change checks:
  - [x] The change is transparent
  - [ ] The change is disruptive
  - [ ] The change requires no migration steps
  - [ ] The change requires migration steps
- Metrics checks:
  - [x] The metrics are still exposed and present in Grafana after the change
  - [x] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [x] The logs do not show any errors after the change
- Network Policy checks:
  - [ ] Any changed pod is covered by Network Policies
  - [ ] The change does not cause any dropped packages in the `NetworkPolicy Dashboard`
- Pod Security Policy checks:
  - [ ] Any changed pod is covered by Pod Security Admission
  - [ ] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [ ] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Falco checks:
  - [ ] The change does not cause any alerts to be generated by Falco
- Audit checks:
  - [x] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy

OlleLarsson commented on WIP-CHANGELOG.md:
```diff
@@ -40,6 +40,7 @@
   - Add pattern matching for node and disk
 - Added support for configuring multiple kube-apiservers for blackbox exporter in sc
 - `./bin/ck8s update-ips` now adds the IPs and port for swift endpoint
+- Added dashboard that showing how spread out pods are across nodes
```
```suggestion
- Added dashboard that shows how spread out pods are across nodes
```

davidumea reviewed (commented):
If I understood the issue correctly, the dashboard should also show zone skew, e.g. by using label `topology.kubernetes.io/zone`. Problem is that **most** of the infra providers that we use at least have all nodes in the same zone. (except e.g. elastx)

Zash commented:
@davidumea 
> Problem is that **most** of the infra providers that we use at least have all nodes in the same zone.

My intent was to do add a second dashboard for zones in a second PR. Are you saying it can be skipped?

davidumea commented:
> My intent was to do add a second dashboard for zones in a second PR.

Okay, I think it would be nice to have it in the same dashboard, that way the user can find all information about how a deployment is spread in one place.

> Are you saying it can be skipped?

No I don't think it should be skipped :slightly_smiling_face:. I meant that maybe we want to do something extra here, possibly sort out/not show the data if there only is one zone in the environment. If more than one zone exists then it should be obvious if multiple replicas of a deployment are in the same zone.

Zash commented:
Pushing an updated version where you can pick between node or zone skew.

Zash commented:
![image](https://github.com/elastisys/compliantkubernetes-apps/assets/197474/b4c924ab-f003-4965-bdf8-d43bfc4395ba)

davidumea reviewed (approved):
This looks really good :+1:

viktor-f commented:
The "pods per node" graph, is that stacked? I think I would prefer if that was not stacked. I think that it might then be easier to see how many pods are on each node.

viktor-f commented:
> > The "pods per node" graph, is that stacked? I think I would prefer if that was not stacked. I think that it might then be easier to see how many pods are on each node.
> 
> Not stacked currently, so lines with the same number of pods overlap.
> 
If they are not stacked, then something seems wrong with the last screenshot you sent. For rook-ceph it is reporting a skew of 2, but looking at the graph I think it should be 6-3=3. Am I missing something?

> > This is shown per deployment/statefulset, right?
>
>  Currently per namespace.
Didn't you have it working for deployment/statefulset earlier? This could still be useful, but it would be much better to see the skew per deployment/statefulset since that is what kubernetes should be balancing.

Zash commented:
> If they are not stacked, then something seems wrong with the last screenshot you sent. For rook-ceph it is reporting a skew of 2, but looking at the graph I think it should be 6-3=3. Am I missing something?

That's shown per node but the view is set to zone granularity, so it's the difference between 8 pods in zone 'a' and 6 pods in zone 'b'. 

> Didn't you have it working for deployment/statefulset earlier? This could still be useful, but it would be much better to see the skew per deployment/statefulset since that is what kubernetes should be balancing.

Sorta, but I had trouble keeping all the labels trough joins with variable granularity. Trying again this morning, I made some progress on that.

Zash commented:
After talking with @viktor-f I'll try to rearrange the per-Namespace graphs to loop over ReplicaSet / StatefulSet instead

Zash commented:
Now with per-ReplicaSet/StatefulSet calculations:

![image](https://github.com/elastisys/compliantkubernetes-apps/assets/197474/58a783fe-a73c-4bf2-9ef5-e29dab7a3969)

OlleLarsson commented on helmfile/upstream/prometheus-community/kube-prometheus-stack/charts/kube-state-metrics/values.yaml:
```diff
@@ -233,7 +233,8 @@ metricDenylist: []
 # severe performance implications (Example: '=pods=[*]').
-metricLabelsAllowlist: []
+metricLabelsAllowlist:
+  - nodes=[topology.kubernetes.io/zone]
```
Isn't it possible to set this in the kube-prometheus-stack values file, e.g. [here](https://github.com/elastisys/compliantkubernetes-apps/blob/main/helmfile/values/kube-prometheus-stack-wc.yaml.gotmpl#L23)?

Zash replied:
Good question. Still new to all these layers of indirection 😄 

Something along these lines perhaps?

``` diff
@@ -21,6 +21,13 @@ prometheusOperator:
     resources: {{- toYaml .Values.prometheusOperator.prometheusConfigReloader.resources | nindent 6 }}
 
 kube-state-metrics:
+
+  # if zones enabled
+  {{- if true }}
+  metricLabelsAllowlist:
+  - nodes=[topology.kubernetes.io/zone]
+  {{- end }}
+
   resources: {{- toYaml .Values.kubeStateMetrics.resources | nindent 4 }}
   {{- if .Values.kubeStateMetrics.clusterAPIMetrics.enabled }}
   rbac:
```

Is there an existing settings for whether zones are used at all yet?

OlleLarsson replied:
> Is there an existing settings for whether zones are used at all yet?

No, not that i'm aware of

Zash replied:
Hm, does it even need to be conditional?

OlleLarsson replied:
Not entirely sure, you could see what happens if the label isn't set on the nodes.

Zash replied:
If it's in the allow-list but no such labels are set it just doesn't show up in the metrics data, so I don't think there'll be any storage overhead for clusters without it. Unless there's some significant optimization for the case of an _empty_ allow-list, I doubt there's much of a performance impact either.

So not adding conditional or new settings for now then.

Zash commented:
Updated first comment with new screenshot. Dropped multi-select for namespaces because it caused wrong results. Dropped the overview for similar reason.

viktor-f reviewed (approved):
LGTM, we can see if we can figure out an overview later.
